metrics.go

```diff
@@ -30,3 +34,61 @@ var adminMetrics = struct {
 	requestCount  *prometheus.CounterVec
 	requestErrors *prometheus.CounterVec
 }{}
+
+// Similar to promhttp.InstrumentHandlerCounter, but upper-cases method names
+// instead of lower-casing them.
+//
+// Unlike promhttp.InstrumentHandlerCounter, this assumes a "code" and "method"
+// label is present, and will panic otherwise.
+func instrumentHandlerCounter(counter *prometheus.CounterVec, next http.Handler) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		d := newDelegator(w)
+		next.ServeHTTP(d, r)
+		counter.With(prometheus.Labels{
+			"code":   sanitizeCode(d.status),
+			"method": sanitizeMethod(r.Method),
+		}).Inc()
+	})
+}
+
+func newDelegator(w http.ResponseWriter) *delegator {
+	return &delegator{
+		ResponseWriter: w,
+	}
+}
+
+type delegator struct {
+	http.ResponseWriter
+	status int
+}
+
+func (d *delegator) WriteHeader(code int) {
+	d.status = code
+	d.ResponseWriter.WriteHeader(code)
+}
+
+func sanitizeCode(s int) string {
+	switch s {
+	case 0, 200:
+		return "200"
+	default:
+		return strconv.Itoa(s)
+	}
+}
+
+func sanitizeMethod(m string) string {
+	switch m {
+	case "GET", "get":
+		return "GET"
+	case "PUT", "put":
+		return "PUT"
+	case "HEAD", "head":
+		return "HEAD"
+	case "POST", "post":
+		return "POST"
+	case "DELETE", "delete":
+		return "DELETE"
+	default:
+		return strings.ToUpper(m)
```

Comment:
Is it actually faster to use a switch rather than just always running `strings.ToUpper`? :thinking:

Reply:
Yes. (o:

Reply:
@francislavoie I had that same question (this code is copied from the Prometheus client), and wrote (then threw away) some benchmarks.

For `sanitizeMethod`, the switch-based benchmark was actually _slightly_ slower (by ~8ns - they're both _around_ 150ns on my laptop), but with fewer allocations. I was a bit surprised about the speed TBH.

For `sanitizeCode`, at least the one in the Prometheus client (I didn't copy it here originally), the switch-based benchmark is about twice as fast time-wise (~120ns for the `if` plus `strconv.Itoa` vs. ~60ns for the switch-based one), and also with fewer allocations.

What I might do (eventually) is pull both the `sanitizeCode` and `sanitizeMethod` functions into an internal package so I don't end up with different implementations for different purposes. But, later. 😉

Reply:
I also just noticed https://github.com/prometheus/client_golang/pull/447#discussion_r225461218 where @beorn7 has effectively explained the same thing, but with a lot more clarity 😉

Reply:
Let's not try and be more clever than the standard library -- just use `strings.ToUpper()`.

Reply:
Ok. We can always go back to something like this if we need to reduce allocations later on.